Update areas

kgoebber opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/Unidata/MetPy/blob/main/CONTRIBUTING.md
-->

#### Description Of Changes

This is a first pass at updating the ways that the areas baked into the declarative syntax works. I've captured 331 areas (a subset of what was in our original list) and have them corresponding to both an extent and a Cartopy projection. This is all contained in a separate file now and removed from the declarative file. These areas, their extents, and the projections are also available outside of the declarative syntax as well, which is captured the tutorial.

In adding elements to the docs I've attempted to capture things in a tutorial as well as a separate listing of all of the areas with a graphical representation of those areas with their long descriptive name. This generates 331 images, so the build of the docs takes longer, but if the file is not changed, then the caching of the images works as intended.

Still to do is to add tests.

Open to suggestions about organization of any of this material.

#### Checklist

- [x] Closes #2526 
- [x] Tests added
- [x] Fully documented

dopplershift commented:
1. Any chance `area_data` was generated so that you could change it easily? I have some performance concerns about creating that many tiny dictionaries at import-time, only to throw them away (vs. e.g. named tuples).
2. I'm not wild about checking in a generated file. Do we really need it committed?

kgoebber commented:
1. It was partially generated - manually did some line wrapping because of character limit, but if I take the time I should be able to do that as well. I can look at using a named tuple instead - initially went with what I knew (e.g., dictionaries) - but I'm not wedded to that if there would be better performance/better data object to use.
2. I tried to have the area.rst generate during the build, but I couldn't seem to find the right incantation of things through all of the sphinx docs settings, etc without a lot of excess stuff being generated as we have set up for a tutorial or gallery page. Reading of the Sphinx documentation was not yielding any clear paths to me from what I was able to find. In the end, I took the inspiration from Cartopy for the generation of their projections list. I would love to have that auto generate instead of committing a generated file.

I'll start looking at a named tuple for creating the areas and if there are any resources you want to point me toward for being able to auto generate the documentation file, I would be very happy to entertain.

Any other thoughts about functionality or organization of the information/documentation?

kgoebber commented on docs/conf.py:
```diff
@@ -12,6 +12,7 @@
 from datetime import datetime
 from pathlib import Path
 import re
+from sphinx_gallery import gen_rst
```
whoops! missed taking this out after testing something...

kgoebber commented:
Okay, so I've updated this PR with what ended up being a very simple way of generating the area.rst file to both make files. I removed the committed area.rst and we should be relying on the autogenerated one from the Python file.

kgoebber commented:
Okay, couldn't find a way with a more specific sphinx way of creating the `areas.rst` file, but I have been able to bake in a method for running the `make_areas.py` file within `conf.py`, which seems more robust than adding it to `Makefile` and `make.bat`.

dopplershift commented:
I've got the best path to integrate with Sphinx here figured out. I'll push some changes to this PR.

dopplershift commented:
So this isn't much different than what was done (running inside `conf.py`), but we can list pretty much any Python file in the same directory as `conf.py` as an extension. That extension then has `setup()`, which registers a callback for an event, which in our case we wait for the "builder" to be initialized.

Other than that it's the same code, though it relies on Sphinx to know some proper paths and give us a logger. Sphinx doesn't have much better in the way of hooks to grab, but this does seem a bit more structured than running a script every time `conf.py` is imported. It's actually the same way Sphinx's own `autosummary` extension is wired in.

dopplershift commented:
@dcamron This will need your review since I've now committed on the PR.

dcamron commented on src/metpy/plots/plot_areas.py:
Would we rather expose `areas` directly in `metpy.plots` if we'll be advocating for its use even outside of declarative?

kgoebber replied:
If there is a way to make it more readily available, that might be worthwhile.

dcamron replied:
@dopplershift any thoughts on the specific names here? Any issues including in `plots.__all__`? I'm glad to make or suggest the change explicitly if we decide to head that direction.

dopplershift replied:
I'm ok with exposing it, but I think it should probably be called something more descriptive than `areas`. Unfortunately, this comment has been open for several hours now and nothing has come to mind.

kgoebber replied:
What about `domain_areas`, `graphics_areas`, `extent_areas`, `defined_areas`, `bounding_areas`, `location_areas`?

dopplershift commented on examples/formats/GINI_Water_Vapor.py:
You shouldn't need to update this example. If you're running into issues, try re-installing from a current main, which should set up the necessary xarray entry point.

dopplershift commented on src/metpy/plots/plot_areas.py:
```diff
@@ -0,0 +1,1014 @@
+
+ccrs = import_cartopy()
+exporter = Exporter(globals())
+Areas = namedtuple('Area', ['area', 'name', 'bounds', 'projection'])
```
How about:
```suggestion
Area = namedtuple('Area', ['name', 'description', 'bounds', 'projection'])
```

dopplershift commented on src/metpy/plots/plot_areas.py:
```diff
@@ -0,0 +1,1014 @@
+exporter = Exporter(globals())
+Areas = namedtuple('Area', ['area', 'name', 'bounds', 'projection'])
+
+area_data = {
+    'us': Areas('us', 'UNITED STATES',
+                (-119.0, -56.0, 19.0, 47.0),
+                ccrs.NorthPolarStereo(central_longitude=-100)),
+    'awips': Areas('awips', 'AWIPS US',
+                   (-127.0, -59.0, 20.0, 50.0),
+                   ccrs.LambertConformal(standard_parallels=[30, 60], central_longitude=-100)),
```
Instead of a dict here, how about just a list for the data to reduce redundancy, and maybe shorten a bunch of these lines...(see below):
```suggestion
area_data = [
    Areas('us', 'UNITED STATES',
          (-119.0, -56.0, 19.0, 47.0),
          ccrs.NorthPolarStereo(central_longitude=-100)),
    Areas('awips', 'AWIPS US',
          (-127.0, -59.0, 20.0, 50.0),
          ccrs.LambertConformal(standard_parallels=[30, 60], central_longitude=-100)),
```

dopplershift commented on src/metpy/plots/plot_areas.py:
```diff
@@ -0,0 +1,1014 @@
+                (25.0, 34.0, -22.9, -15.5),
+                ccrs.PlateCarree())}
+
+with exporter:
+    areas = area_data
```
Then generate the dict here and we avoid having the "name" in two spots:
```suggestion
with exporter:
    named_areas = {data.name: data for data in area_data}
```

dopplershift reviewed (changes requested):
Sorry, finally noticed a few things here.

If you can't get to them before (trying to) release tomorrow, I'll take care of them.

kgoebber commented:
I was able to make the requested changes, but don't have the correct method of exposing the `named_areas` from `plots` directly and not run afoul of import things. I think you all will have to take it from here.

dopplershift commented:
I confirmed that the default to dynamically control whether we generate ALL the areas is working properly on CI vs. locally. This is a significant time savings for building locally (when you don't actually care about having them all).
